v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+                {
+                    break;
+                }
+                else
+                {
+                    pos++;
+                }
+            }
+        }
+        else
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_082: [ If more than 64 properties are given in log_context, log_sink_etw.log_sink_log shall not add any properties to the event. ]*/
+            add_properties = false;
+        }
+
+        /* Codes_SRS_LOG_SINK_ETW_01_026: [ log_sink_etw.log_sink_log shall fill a SELF_DESCRIBED_EVENT structure, setting the following fields: ]*/
+        uint16_t metadata_size;
+
+        if (!add_properties)
+        {
+            metadata_size = metadata_size_without_properties;
+            pos = &self_described_event->metadata[metadata_size_without_properties];
+        }
+        else
+        {
+            metadata_size = (uint16_t)(pos - &self_described_event->metadata[0]);
+        }
+
+        char* formatted_message = (char*)pos;
+        size_t available_bytes = sizeof(_tlgEvent) - (formatted_message - (char*)self_described_event);
+
+        int formatted_message_length = vsnprintf(formatted_message, available_bytes, message, args);
+        if (formatted_message_length < 0)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_086: [ If any error occurs log_sink_etw.log_sink_log shall print Error emitting ETW event and return. ]*/
+            (void)printf("Error emitting ETW event.\r\n");
+        }
+        else
+        {
+            if ((size_t)formatted_message_length >= available_bytes)
+            {
+                if (add_properties)
+                {
+                    // did not fit, print again without properties and limit it
+                    /* Codes_SRS_LOG_SINK_ETW_01_085: [ If the size of the metadata and the formatted message exceeds 4096 bytes, log_sink_etw.log_sink_log shall not add any properties to the event. ]*/
+                    add_properties = false;
+                    metadata_size = metadata_size_without_properties;
+
+                    formatted_message = (char*)&self_described_event->metadata[metadata_size_without_properties];
+                    available_bytes = sizeof(_tlgEvent) - (formatted_message - (char*)self_described_event);
+                    formatted_message_length = vsnprintf(formatted_message, available_bytes, message, args);
+                    if (formatted_message_length < 0)
+                    {
+                        /* Codes_SRS_LOG_SINK_ETW_01_086: [ If any error occurs log_sink_etw.log_sink_log shall print Error emitting ETW event and return. ]*/
+                        (void)printf("Error emitting ETW event.\r\n");
+                    }
+                    else
+                    {
+                        // Ok as is
+                    }
+                }
+            }
+            else
+            {
+                // Ok as is
+            }
+
+            if (formatted_message_length >= 0)
```

Comment:
>>= [](http://example.com/codeflow?start=41&length=2)

Can this condition ever be "false"?

Reply:
Yes, it can, if we attempt to printf the message in line 293 and it fails with a negative return then we can't do much more about it.